Add map camera functionality, better config flow, and initial testing coverage

prairiesnpr opened the pull request:
- Adds support for setting a unique map for each mower. Resolves #409 
- Adds support for showing multiple mowers on a single map.  Resolves #361, resolves #364
- Add support for specifying a rotation, for a map to allow for images that aren't oriented North and South. Resolves #462, Resolves #441
- Config flow enforces proper data to reduce support questions
- Adds initial unit testing

[Coverage Report](https://app.codecov.io/gh/prairiesnpr/husqvarna_automower/tree/dev/) [![codecov](https://codecov.io/gh/prairiesnpr/husqvarna_automower/branch/dev/graph/badge.svg?token=L3T457D0WJ)](https://codecov.io/gh/prairiesnpr/husqvarna_automower)

**Two Mower Map, With Rotation**
![mower_one_out](https://github.com/Thomas55555/husqvarna_automower/assets/18637008/e4e69a88-e9bb-487d-80f0-67c6af769efc)

**Single Mower Map, With Rotation**
![mower_two_out](https://github.com/Thomas55555/husqvarna_automower/assets/18637008/db68460e-779b-4703-9a70-3be3814c0b4b)

Thomas55555 commented:
Okay, that's I huge PR. I try to review it.
But first a question. How do I run the tests? I've never done this before. I've tried this:
```
(venv) @Thomas55555 ➜ /workspaces/husqvarna_automower/custom_components/husqvarna_automower (pr/prairiesnpr/485) $ pytest
ImportError while loading conftest '/workspaces/husqvarna_automower/custom_components/husqvarna_automower/tests/conftest.py'.
__init__.py:6: in <module>
    from homeassistant.components.application_credentials import DATA_STORAGE
../../../core/homeassistant/components/application_credentials/__init__.py:16: in <module>
    from homeassistant.components import websocket_api
../../../core/homeassistant/components/websocket_api/__init__.py:13: in <module>
    from . import commands, connection, const, decorators, http, messages  # noqa: F401
../../../core/homeassistant/components/websocket_api/commands.py:49: in <module>
    from . import const, decorators, messages
../../../core/homeassistant/components/websocket_api/decorators.py:15: in <module>
    from .connection import ActiveConnection
../../../core/homeassistant/components/websocket_api/connection.py:13: in <module>
    from homeassistant.components.http import current_request
../../../core/homeassistant/components/http/__init__.py:42: in <module>
    from .ban import setup_bans
../../../core/homeassistant/components/http/ban.py:18: in <module>
    from homeassistant.components import persistent_notification
../../../core/homeassistant/components/persistent_notification/__init__.py:206: in <module>
    @websocket_api.websocket_command({vol.Required("type"): "persistent_notification/get"})
E   AttributeError: partially initialized module 'homeassistant.components.websocket_api' has no attribute 'websocket_command' (most likely due to a circular import)
``

prairiesnpr commented:
If you decide to merge it, I have the github action setup to upload the coverage results, you would need to register the repo with codecov, [codecov.com](https://docs.codecov.com/docs).  It will still pass, just will give a warning that it wasn't able to upload the results.

Thomas55555 commented:
Hi, I started reviewing and also testing.
First thing was, the zones are here now(still thing it's not right here)... But I gave it a try.
When trying to configure it. There was an error, that at least three coordinates are required. But I entered 3. But didn't work. After entering 4 coordinates, it worked. I could also "save" the zone. But when I assigned it to the mower. I think I have to assign it by clicking on the button of the mower name. It dissapears.

Thomas55555 commented:
Ah okay. I have to click on "Save zone" first. Didn't got this. I assumed it's already saved, after it's created.

Thomas55555 commented:
It was even in the documentation, that I have to save it :D
Okay, no I saved a home zone. Assigned it to the mower. If I look on the map, I see the zone, but after clicking on the map again, the map is not available anymore.

prairiesnpr commented:
Good catch, I'll update the options flow to verify that we ensure a mower was added and that we have a test for it to confirm it's all working.

Thomas55555 commented:
> I should note this also will resize the camera image before passing it back to HA, seems to be a little faster now. In addition, this would restore the ability to configure low energy mode. That option was removed after I had already reworked the configuration. It's working as advertised, so I would vote to leave it in, but if we want to remove it, I can do that.

Yes, it was working. But the problem is, the mower isn't sending en error message when it's in an error mode on the websocket. This only happens when polling. But if we poll only once a day, we have to wait a long time to see, that the mower has an error. So I think it's better, when it's gone again.

prairiesnpr commented:
> Hi, I started reviewing and also testing. First thing was, the zones are here now(still thing it's not right here)... But I gave it a try. When trying to configure it. There was an error, that at least three coordinates are required. But I entered 3. But didn't work. After entering 4 coordinates, it worked. I could also "save" the zone. But when I assigned it to the mower. I think I have to assign it by clicking on the button of the mower name. It dissapears.

I understand your point on the zone sensor, I would like to keep it for a couple reasons. One, ha isn't going to include polygon zones due to limitations in android/iOS and second, this would sync our codebases, which would allow future PRs to be simpler.

Thomas55555 commented:
> Good catch, I'll update the options flow to verify that we ensure a mower was added and that we have a test for it to confirm it's all working.

And the map is still not working anymore after I clicked on it once, and the zone is gone after that. You should also have a look for that.

prairiesnpr commented:
> > I should note this also will resize the camera image before passing it back to HA, seems to be a little faster now. In addition, this would restore the ability to configure low energy mode. That option was removed after I had already reworked the configuration. It's working as advertised, so I would vote to leave it in, but if we want to remove it, I can do that.
> 
> Yes, it was working. But the problem is, the mower isn't sending en error message when it's in an error mode on the websocket. This only happens when polling. But if we poll only once a day, we have to wait a long time to see, that the mower has an error. So I think it's better, when it's gone again.

Should we make it a configurable pull rate, maybe allow a choice in seconds?

prairiesnpr commented:
> > 
> 
> I also thought about that, or even dynamic polling rates. But I think it's better to keep it as simple as possible for the user and just poll every 300s.

Cool, I'll remove that logic, it will take just a bit, have to travel for work tomorrow, may not be able to get to it today.

prairiesnpr commented:
> > Good catch, I'll update the options flow to verify that we ensure a mower was added and that we have a test for it to confirm it's all working.
> 
> And the map is still not working anymore after I clicked on it once, and the zone is gone after that. You should also have a look for that.
Can you give me the exact steps to replicate this, just want to make sure I solve it, didn't happen to me when I was testing.

Thomas55555 commented on custom_components/husqvarna_automower/device_tracker.py:
```diff
@@ -1,56 +1,56 @@
+) -> None:
+    """Set up device_tracker platform."""
+    coordinator = hass.data[DOMAIN][entry.entry_id]
+    entity_list = []
+    for idx, ent in enumerate(coordinator.session.data["data"]):
+        try:
+            coordinator.session.data["data"][idx]["attributes"]["positions"][0][
+                "latitude"
+            ]
+            entity_list.append(AutomowerTracker(coordinator, idx))
+        except IndexError:
+            pass
+
+    async_add_entities(entity_list)
+
+
+class AutomowerTracker(TrackerEntity, AutomowerEntity):
+    """Defining the Device Tracker Entity."""
+
+    def __init__(self, session, idx):
+        """Initialize AutomowerDeviceTracker."""
+        super().__init__(session, idx)
+        self._attr_unique_id = f"{self.mower_id}_dt"
+
+    @property
+    def source_type(self) -> SourceType:
+        """Return the source type, eg gps or router, of the device."""
+        return SourceType.GPS
+
+    @property
+    def latitude(self) -> float:
+        """Return latitude value of the device."""
+        lat = AutomowerEntity.get_mower_attributes(self)["positions"][0]["latitude"]
+        return lat
+
+    @property
+    def longitude(self) -> float:
+        """Return longitude value of the device."""
+        lon = AutomowerEntity.get_mower_attributes(self)["positions"][0]["longitude"]
+        return lon
```
Why is this changed completely?

Thomas55555 commented on custom_components/husqvarna_automower/manifest.json:
```diff
@@ -1,25 +1,26 @@
-  ],
-  "version": "0.0.0"
-}
+{
+  "domain": "husqvarna_automower",
+  "name": "Husqvarna Automower",
+  "codeowners": [
+    "@Thomas55555"
+  ],
+  "config_flow": true,
+  "dependencies": [
+    "application_credentials",
+    "my",
+    "schedule"
+  ],
+  "documentation": "https://github.com/Thomas55555/husqvarna_automower",
+  "integration_type": "hub",
+  "iot_class": "cloud_push",
+  "issue_tracker": "https://github.com/Thomas55555/husqvarna_automower/issues",
+  "loggers": ["aioautomower", "geopy", "numpy", "Pillow"],
+  "requirements": [
+    "aioautomower==2023.4.4",
+    "geopy>=2.1.0",
+    "numpy>=1.21.6",
+    "Pillow>=9.1.1",
+    "Shapely>=1.8.2"
+  ],
+  "version": "0.0.0"
+}
```
Why is this changed completely?

Thomas55555 commented on custom_components/husqvarna_automower/translations/sv.json:
```diff
@@ -1,61 +1,61 @@
+                "description": "Aktivera kartkamera",
+                "title": "Husqvarna Automower-inställningar"
+            },
+            "config": {
+                "data": {
+                    "gps_top_left": "GPS-koordinater för det övre vänstra bildhörnet",
+                    "gps_bottom_right": "GPS-koordinater för det nedre högra bildhörnet",
+                    "mower_img_path": "Sökväg till bild på gräsklipparen",
+                    "map_img_path": "Sökväg till kartbilden"
+
+                },
+                "description": "Kamerainställningar",
+                "title": "Husqvarna Automower-inställningar"
+            }
+        }
+    },
+    "system_health": {
+        "info": {
+            "can_reach_server": "Husqvarna Token-server kan nås"
+        }
+    },
+    "application_credentials": {
+      "description": "Skapa ett konto på [Husqvarna developers portal]({oauth_creds_url}). Anslut Authentication API and Automower Connect API. Som redirect-URI anges https://my.home-assistant.io/redirect/oauth. Om du inte har konfigurerat My HomeAssistant, ange istället denna URL: {redirect_uri}"
+    },
+    "entity": {
+        "vacuum": {
+            "quirk": {
+                "state": {
+                    "cleaning": "Klipper"
+                }
+            }
+        },
+        "camera": {
+            "quirk": {
+                "state": {
+                    "cleaning": "Klipper"
+                }
+            }
+        }
+    }
```
Why is this changed completely?

Thomas55555 reviewed (changes requested):
Okay, here are the first changes I suggest. I hope I can extend the review, when I discover something else.
What I really like, are the tests. The make work easier I hope.

prairiesnpr commented:
I think the files showing complete changes is a line ending issue.  See https://stackoverflow.com/questions/30752959/git-showing-entire-file-is-modified-instead-of-showing-modified-small-part-of-co

prairiesnpr commented on custom_components/husqvarna_automower/sensor.py:
```diff
@@ -258,6 +264,7 @@ async def async_setup_entry(
     coordinator = hass.data[DOMAIN][entry.entry_id]
     entity_list = []
     for idx, ent in enumerate(coordinator.session.data["data"]):
+        entity_list.append(AutomowerZoneSensor(coordinator, idx, entry))
```
Just to make sure I'm clear, you don't want to have the sensor disabled via ```_attr_entity_registry_enabled_default``` you don't want the sensor to be added at all in the case we don't have any zones created?

prairiesnpr replied:
done, changed to mimic camera

prairiesnpr commented on custom_components/husqvarna_automower/__init__.py:
```diff
@@ -1,111 +1,118 @@
+        _LOGGER.info(STARTUP_MESSAGE)
+    coordinator = AutomowerDataUpdateCoordinator(
+        hass,
+        entry=entry,
+    )
+    await coordinator.async_config_entry_first_refresh()
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+    entry.async_on_unload(entry.add_update_listener(update_listener))
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Handle unload of an entry."""
+    coordinator = hass.data[DOMAIN][entry.entry_id]
+    await coordinator.session.close()
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+    return unload_ok
+
+
+async def async_reload_entry(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Reload config entry."""
+    await hass.config_entries.async_reload(entry.entry_id)
+
+
+async def update_listener(
+    hass: HomeAssistant, entry: ConfigEntry
+) -> None:  # Todo: Add test
+    """Handle options update."""
+    unload_ok = await hass.config_entries.async_unload_platforms(
+        entry, [Platform.CAMERA]
+    )
+    if unload_ok:
+        await hass.config_entries.async_forward_entry_setups(entry, [Platform.CAMERA])
+        entry.async_on_unload(entry.add_update_listener(update_listener))
```
Still convinced this a line ending issue, should be good once we standardize on LF

prairiesnpr commented:
I still have to add the translation key for the zone sensor and related json entries.  Also seeing testing errors on all of the sensors, it fails an assert when repr is called on the sensor object.  Need to understand if this is just an error with the testing framework or if we are missing something, may be related to the translation PR that was just merged, don't know, will look into it further when I return this weekend.

prairiesnpr commented:
@Thomas55555  This should now address your comments.  

The previously mentioned issue with translations is isolated to the test framework, they are working properly.
I did remove the core file that was added, seemed like an oversight.

Other than that, tests all pass, also ran in a current core dev container and everything worked well.  Let me know if you have any other thoughts/questions

Din-BH commented:
> > 
> 
> I also thought about that, or even dynamic polling rates. But I think it's better to keep it as simple as possible for the user and just poll every 300s.

It seems to me that updates to the map (positions) are less frequent than in the app, although have not vwrified it lately. Are the coordinate changes supposed to come via the websock or is that a polling quesrtion? If it is about polling then I for one would maybe suggest different poll rates for two different scenarios. When the mower is mowing, I would poll more frequently (more likely that error occors then, and even more so if it would increase map reliability) and then less frequently - maybe then the 300s - when not mowing. It saams there are some issues getting some other states properly without polling. Need to look at the API a bit closer to verify my thoughts.

World not mind these two being possible to configure for advanced users or for testing. Is there a limit for API use set by Husqvarna? If so then that needs to be factored in so we don't start getting such issues..

prairiesnpr commented:
> > > 
> > 
> > 
> > I also thought about that, or even dynamic polling rates. But I think it's better to keep it as simple as possible for the user and just poll every 300s.
> 
> It seems to me that updates to the map (positions) are less frequent than in the app, although have not vwrified it lately. Are the coordinate changes supposed to come via the websock or is that a polling quesrtion? If it is about polling then I for one would maybe suggest different poll rates for two different scenarios. When the mower is mowing, I would poll more frequently (more likely that error occors then, and even more so if it would increase map reliability) and then less frequently - maybe then the 300s - when not mowing. It saams there are some issues getting some other states properly without polling. Need to look at the API a bit closer to verify my thoughts.
> 
> World not mind these two being possible to configure for advanced users or for testing. Is there a limit for API use set by Husqvarna? If so then that needs to be factored in so we don't start getting such issues..

See https://github.com/Thomas55555/husqvarna_automower/discussions/299 for some research we did last year on the same topic.

prairiesnpr commented on custom_components/husqvarna_automower/camera.py:
```diff
@@ -120,21 +211,79 @@ def supported_features(self) -> int:
-            self._position_history = position_history + self._position_history
-            position_history = self._position_history
+            self._position_history[mower_id] = (
+                position_history + self._position_history[mower_id]
```
Is this a fresh install or was there an existing config entry?  I'll catch that error and add a test for it, but its odd to get an update but no existing position history for that mower.

Thomas55555 replied:
Existing condig, but I think it's really the quotation marks around `mower_id`?

prairiesnpr replied:
mower_id is a variable passed in to the function l, I have to track each position by mower Id now so we can have multiple mowers on a map, I'm wondering if since it's an existing config, that's not present, it's an easy fix, but a great call out also.  I'm testing on a new config, so never saw it.

prairiesnpr replied:
Latest commit fixes this, caused by getting a single coordinate position update with no previous position history.  We now account for it and there is a test added to replicate.

Din-BH commented:
Not sure if I should have uninstalled something first, but I just renamed the original husqvarna_automover to get back to later and copied in your code to the directory instead. The integration seemed to work as is - as expected, but when I tried to configure, I am getting this

![Husq_prairiesnpr_config](https://github.com/Thomas55555/husqvarna_automower/assets/58572670/87804a68-0959-4b31-88fe-f8909ccbb348)

Thomas55555 commented:
And another thing, if I change a zone, the selection `Show zone on map camera` is remembered by the integration, but the assigned mowers are not remembered, or at least I can't see, that the corresponding field in the config flow is blue with a check. Can you add this?

Thomas55555 commented:
> 

Ctrl + F5?

prairiesnpr commented:
> > > Not sure if I should have uninstalled something first, but I just renamed the original husqvarna_automover to get back to later and copied in your code to the directory instead. The integration seemed to work as is - as expected, but when I tried to configure, I am getting this
> > > ![Husq_prairiesnpr_config](https://user-images.githubusercontent.com/58572670/246181715-87804a68-0959-4b31-88fe-f8909ccbb348.jpg)
> > 
> > 
> > Ctrl + F5 ?
> 
> Not helping...
> 
> Also next step is not obvious - seems we are missing some texts
> 
> ![Husq_prairiesnpr_config2](https://user-images.githubusercontent.com/58572670/246183469-e42515ae-75d3-4d49-b845-f2203cad9c13.jpg)

That doesn't look like the latest form?  Think you may have the old version still. It should look like this, before we removed the le config

Camera config
![Screenshot_20230615_125834_Edge](https://github.com/Thomas55555/husqvarna_automower/assets/18637008/d95bcebf-db22-46c9-bed0-535beea1294f)
![Screenshot_20230615_125842_Edge](https://github.com/Thomas55555/husqvarna_automower/assets/18637008/c2a15851-a716-42dd-a0b4-933ac968f334)
![Screenshot_20230615_125849_Edge](https://github.com/Thomas55555/husqvarna_automower/assets/18637008/6d789662-7fe8-47ed-bddb-12c83a392459)

Zone config 
![Screenshot_20230615_125857_Edge](https://github.com/Thomas55555/husqvarna_automower/assets/18637008/37bf0792-8dfd-4d13-9c55-4fd2dd39696a)
![Screenshot_20230615_125905_Edge](https://github.com/Thomas55555/husqvarna_automower/assets/18637008/22aad9ed-92c6-4aef-b8c4-5dcdd5f62297)

Din-BH commented:
Hmm. there were no further options after that second screen, which was odd indeed. Do I need to specify something specific to get the newest when retrieving the code from your git?

prairiesnpr commented:
> Hmm. there were no further options after that second screen, which was odd indeed. Do I need to specify something specific to get the newest when retrieving the code from your git?

If you are pulling from github, make sure you get dev, not main, also, makes sure you restart home assistant.  That's the two most likely issues that I can think of.

Din-BH commented:
> > Hmm. there were no further options after that second screen, which was odd indeed. Do I need to specify something specific to get the newest when retrieving the code from your git?
> 
> If you are pulling from github, make sure you get dev, not main, also, makes sure you restart home assistant. That's the two most likely issues that I can think of.

Ok. I did restart, but did not specify branch. I will try again.

prairiesnpr commented:
> And another thing, if I change a zone, the selection `Show zone on map camera` is remembered by the integration, but the assigned mowers are not remembered, or at least I can't see, that the corresponding field in the config flow is blue with a check. Can you add this?

After more research, this doesn't appear to be possible, there is no provision in the selector source to set an initial or default value. See https://github.com/home-assistant/core/blob/5e55f83cbc769080ea8acc78c142233aa49deaeb/homeassistant/helpers/selector.py#L933

prairiesnpr commented:
The last comment will bump the config version to 3 as well as add a migration to the new options format.  I have a test for this and as a sanity check I installed the current version in a new dev container, then upgraded to this PR to ensure the options migrated correctly.  That said, if you use it in a test environment where you have already installed this PR, it may overwrite some options during the migration.

prairiesnpr commented:
Some videos of the new options flow for those who may not want to test.

### Configuring the camera:

https://github.com/Thomas55555/husqvarna_automower/assets/18637008/e73c0b00-403a-49dc-b34c-f808ee1cc746

### Adding a new zone:

https://github.com/Thomas55555/husqvarna_automower/assets/18637008/dfd936de-9840-4548-b89f-52caaea0fd61

Din-BH commented:
Ok. Did a complete restore, unpacked the latest dev - now to the right place in the directory, and it fit like a glove! Rotation seems to be working, but getting the coordinates right is a bit complicated. How about a switch to paint all coordinates on the image? That would make it a *lot* easier to debug the coordinates. I was just about to test if there was already something like this implemented in the Zones menu when...

During my coordinate tweaking (several times changing coordinates back and forth) I seem to have found some discrepancy. Trying to enter CONFIGURE flow now returns Internal error 500.

```
This error originated from a custom integration.

Logger: homeassistant
Source: custom_components/husqvarna_automower/entity.py:38
Integration: Husqvarna Automower (documentation, issues)
First occurred: 09:18:57 (45 occurrences)
Last logged: 09:18:58

Error doing job: Exception in callback AutomowerCamera.__init__.<locals>.<lambda>({'errors': [{'code': 'invalid.credentials', 'detail': 'The supplied... are invalid.', 'id': 'f4e11f4c-922...-eaf2a5f632ab', 'status': '403', ...}]}) at /config/custom_components/husqvarna_automower/camera.py:102
Error doing job: Exception in callback AutomowerEntity.async_added_to_hass.<locals>.<lambda>({'errors': [{'code': 'invalid.credentials', 'detail': 'The supplied... are invalid.', 'id': 'f4e11f4c-922...-eaf2a5f632ab', 'status': '403', ...}]}) at /config/custom_components/husqvarna_automower/entity.py:53
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/config/custom_components/husqvarna_automower/camera.py", line 102, in <lambda>
    lambda data: self._generate_image(data), schedule_immediately=True
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/husqvarna_automower/camera.py", line 341, in _generate_image
    position_history = AutomowerEntity.get_mower_attributes(self)["positions"]
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/husqvarna_automower/entity.py", line 38, in get_mower_attributes
    return self.coordinator.session.data["data"][self.idx]["attributes"]
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^
KeyError: 'data'

```

then

```
Logger: aioautomower.rest
Source: runner.py:179
First occurred: 09:18:57 (2 occurrences)
Last logged: 09:18:59

Response mower data: {'errors': [{'id': 'f4e11f4c-922b-444f-b1c4-eaf2a5f632ab', 'status': '403', 'code': 'invalid.credentials', 'title': 'Invalid credentials', 'detail': 'The supplied credentials are invalid.'}]}
Response mower data: {'errors': [{'id': 'e720108d-d2c2-4402-afc2-e2c4c9cdef0c', 'status': '403', 'code': 'invalid.credentials', 'title': 'Invalid credentials', 'detail': 'The supplied credentials are invalid.'}]}
```

and then

```
This error originated from a custom integration.

Logger: aiohttp.server
Source: custom_components/husqvarna_automower/config_flow.py:126
Integration: Husqvarna Automower (documentation, issues)
First occurred: 09:21:24 (1 occurrences)
Last logged: 09:21:24

Error handling request
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/aiohttp/web_protocol.py", line 433, in _handle_request
    resp = await request_handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aiohttp/web_app.py", line 504, in _handle
    resp = await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aiohttp/web_middlewares.py", line 117, in impl
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/security_filter.py", line 85, in security_filter_middleware
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/forwarded.py", line 100, in forwarded_middleware
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/request_context.py", line 28, in request_context_middleware
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/ban.py", line 80, in ban_middleware
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/auth.py", line 236, in auth_middleware
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/view.py", line 148, in handle
    result = await handler(request, **request.match_info)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/config/config_entries.py", line 219, in post
    return await super().post(request)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/data_validator.py", line 72, in wrapper
    result = await method(view, request, data, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/data_entry_flow.py", line 71, in post
    result = await self._flow_mgr.async_init(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 233, in async_init
    flow = await self.async_create_flow(handler, context=context, data=data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 1830, in async_create_flow
    return HANDLERS[entry.domain].async_get_options_flow(entry)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/husqvarna_automower/config_flow.py", line 110, in async_get_options_flow
    return OptionsFlowHandler(config_entry)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/husqvarna_automower/config_flow.py", line 126, in __init__
    self.hass.data[DOMAIN][entity].session.data["data"]
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^
KeyError: 'data'

```

Din-BH commented:
Another suggestion; When you enter too few coordinates and try to submit a zone, it would be nice if you would be able to add the additional info with the old values intact. Now all values are cleared on error, so you need to start inputting everything again from scratch.

Din-BH commented:
When rotation is changed, zones disappear.

prairiesnpr commented:
> Ok. Did a complete restore, unpacked the latest dev - now to the right place in the directory, and it fit like a glove! Rotation seems to be working, but getting the coordinates right is a bit complicated. How about a switch to paint all coordinates on the image? That would make it a _lot_ easier to debug the coordinates. I was just about to test if there was already something like this implemented in the Zones menu when...
> 
> During my coordinate tweaking (several times changing coordinates back and forth) I seem to have found some discrepancy. Trying to enter CONFIGURE flow now returns Internal error 500.
> 
> ```
> This error originated from a custom integration.
> 
> Logger: homeassistant
> Source: custom_components/husqvarna_automower/entity.py:38
> Integration: Husqvarna Automower (documentation, issues)
> First occurred: 09:18:57 (45 occurrences)
> Last logged: 09:18:58
> 
> Error doing job: Exception in callback AutomowerCamera.__init__.<locals>.<lambda>({'errors': [{'code': 'invalid.credentials', 'detail': 'The supplied... are invalid.', 'id': 'f4e11f4c-922...-eaf2a5f632ab', 'status': '403', ...}]}) at /config/custom_components/husqvarna_automower/camera.py:102
> Error doing job: Exception in callback AutomowerEntity.async_added_to_hass.<locals>.<lambda>({'errors': [{'code': 'invalid.credentials', 'detail': 'The supplied... are invalid.', 'id': 'f4e11f4c-922...-eaf2a5f632ab', 'status': '403', ...}]}) at /config/custom_components/husqvarna_automower/entity.py:53
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.11/asyncio/events.py", line 80, in _run
>     self._context.run(self._callback, *self._args)
>   File "/config/custom_components/husqvarna_automower/camera.py", line 102, in <lambda>
>     lambda data: self._generate_image(data), schedule_immediately=True
>                  ^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/config/custom_components/husqvarna_automower/camera.py", line 341, in _generate_image
>     position_history = AutomowerEntity.get_mower_attributes(self)["positions"]
>                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/config/custom_components/husqvarna_automower/entity.py", line 38, in get_mower_attributes
>     return self.coordinator.session.data["data"][self.idx]["attributes"]
>            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^
> KeyError: 'data'
> ```
> 
> then
> 
> ```
> Logger: aioautomower.rest
> Source: runner.py:179
> First occurred: 09:18:57 (2 occurrences)
> Last logged: 09:18:59
> 
> Response mower data: {'errors': [{'id': 'f4e11f4c-922b-....-eaf2a5f632ab', 'status': '403', 'code': 'invalid.credentials', 'title': 'Invalid credentials', 'detail': 'The supplied credentials are invalid.'}]}
> Response mower data: {'errors': [{'id': 'e720108d-d2c2-....-e2c4c9cdef0c', 'status': '403', 'code': 'invalid.credentials', 'title': 'Invalid credentials', 'detail': 'The supplied credentials are invalid.'}]}
> ```
> 
> and then
> 
> ```
> This error originated from a custom integration.
> 
> Logger: aiohttp.server
> Source: custom_components/husqvarna_automower/config_flow.py:126
> Integration: Husqvarna Automower (documentation, issues)
> First occurred: 09:21:24 (1 occurrences)
> Last logged: 09:21:24
> 
> Error handling request
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.11/site-packages/aiohttp/web_protocol.py", line 433, in _handle_request
>     resp = await request_handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/aiohttp/web_app.py", line 504, in _handle
>     resp = await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/aiohttp/web_middlewares.py", line 117, in impl
>     return await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/src/homeassistant/homeassistant/components/http/security_filter.py", line 85, in security_filter_middleware
>     return await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/src/homeassistant/homeassistant/components/http/forwarded.py", line 100, in forwarded_middleware
>     return await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/src/homeassistant/homeassistant/components/http/request_context.py", line 28, in request_context_middleware
>     return await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/src/homeassistant/homeassistant/components/http/ban.py", line 80, in ban_middleware
>     return await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/src/homeassistant/homeassistant/components/http/auth.py", line 236, in auth_middleware
>     return await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/src/homeassistant/homeassistant/components/http/view.py", line 148, in handle
>     result = await handler(request, **request.match_info)
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/src/homeassistant/homeassistant/components/config/config_entries.py", line 219, in post
>     return await super().post(request)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/src/homeassistant/homeassistant/components/http/data_validator.py", line 72, in wrapper
>     result = await method(view, request, data, *args, **kwargs)
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/src/homeassistant/homeassistant/helpers/data_entry_flow.py", line 71, in post
>     result = await self._flow_mgr.async_init(
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 233, in async_init
>     flow = await self.async_create_flow(handler, context=context, data=data)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/src/homeassistant/homeassistant/config_entries.py", line 1830, in async_create_flow
>     return HANDLERS[entry.domain].async_get_options_flow(entry)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/config/custom_components/husqvarna_automower/config_flow.py", line 110, in async_get_options_flow
>     return OptionsFlowHandler(config_entry)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/config/custom_components/husqvarna_automower/config_flow.py", line 126, in __init__
>     self.hass.data[DOMAIN][entity].session.data["data"]
>     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^
> KeyError: 'data'
> ```

Looks like the Husqvarna api chocked while you were testing,  but we cna certainly handle the api failure more gracefully, I'll add a check to ensure data is present.

prairiesnpr commented:
> Another suggestion; When you enter too few coordinates and try to submit a zone, it would be nice if you would be able to add the additional info with the old values intact. Now all values are cleared on error, so you need to start inputting everything again from scratch.

Good thought, I wasn't certain which would make more sense, should be an easy change to keep the original failed entry.

prairiesnpr commented:
> When rotation is changed, zones disappear.

Did you reload the integration?  I cache the image with the zones so we don't redraw them each time.  May need to force a reload if the values are updated.

prairiesnpr commented:
> > When rotation is changed, zones disappear.
> 
> Did you reload the integration? I cache the image with the zones so we don't redraw them each time. May need to force a reload if the values are updated.

Ignore this thought, it's dumping the zone config when you edit a camera config, looking into it

prairiesnpr commented:
@Din-BH  Latest commit addresses the bug where it would drop the zones if you made changes to the camera config as well as passing the invalid data back to the zone form if there is an error.

As far as the api credential error, trying to catch that is going to be fairly complex, just catching the exception there would push it to another line, we would need to make significant changes to the coordinator to remove all the hard coded keys to make that work.  This PR is already huge, I vote we leave it as is and address in another PR.  This PR doesn't change any of the get_mower_attribute logic, so it's not a regression from the current code base.

As far as drawing the coordinates on the map image, that should be possible, maybe a debug setting?  Any chance you could mock an image up with what you think that could look like and I'll give it some thought.

Din-BH commented:
@prairiesnpr OK. I did a restart on the error. That did not fully recover it, but it seems some parts are working, some not. So I get cutting size and positions, but not status. Quite odd. Anyway, have been doing lots of config after that, and same issue is not there now, so I agree may be good to leave for now.

On the map coordinates display - I have been playing with the zone now to try to get the coordinates right. I am thinking it could be as simple as what a zone looks like. Maybe with at "cross over" (corner to corner) to make it easier to see corners. Setting up coordinates the way it is not is quite a lot of work and this could help make it faster. Yes; I think a debug options sounds good - maybe "Debug" does not really explain enough however, so maybe "Debug map area" toggle?

On a sidenote, the having to save the zones after they are created/changed is super-unintuitive unfortunately. I doubt that people will understand it.

prairiesnpr commented:
> On a sidenote, the having to save the zones after they are created/changed is super-unintuitive unfortunately. I doubt that people will understand it.

I agree, it's far from ideal.  I haven't really come up with a better solution though.  I'll take a second look and see if I can come up with an alternative, the options flows forms are really limiting when it comes to things like this. Maybe we could modify the description of the form to say that you must choose save zones to save your changes?

Din-BH commented:
Hmm. Maybe you need to check the logic of prompting for re-authenticating again? When I re-installed the main version from backup, Thomas' code prompted med to reconfigure the extension. With your version that did not happen?

prairiesnpr commented:
> Hmm. Maybe you need to check the logic of prompting for re-authenticating again? When I re-installed the main version from backup, Thomas' code prompted med to reconfigure the extension. With your version that did not happen?

I'll look again, but shouldn't be any changes there, that's not something I touched.

prairiesnpr commented:
> Is there no way to make the change to the zone happen when you close the zone dialogue?

I can, but it ends the flow, so you would need to restart the entire process to add a new zone, which may be clearer?  I just found it annoying and a bit jarring.

Din-BH commented:
``

> > Is there no way to make the change to the zone happen when you close the zone dialogue?
> 
> I can, but it ends the flow, so you would need to restart the entire process to add a new zone, which may be clearer? I just found it annoying and a bit jarring.

True. But I think that will still be the best option at the end.

prairiesnpr commented:
> ``
> 
> > > Is there no way to make the change to the zone happen when you close the zone dialogue?
> > 
> > 
> > I can, but it ends the flow, so you would need to restart the entire process to add a new zone, which may be clearer? I just found it annoying and a bit jarring.
> 
> True. But I think that will still be the best option at the end.

OK, that's a simple change, I'll try tonight and you can try it and see if that works better for you.  Other than that, any concerns? Always good to get real world feedback.

Din-BH commented:
Getting the coordinates right on the rotated image is quite a struggle.. I think some kind of documentation of intended workflow would be good. I have now set something up - will know more on how good it is tomorrow when the mower gets going again, but to be honest even taking help from my wife who works with maps and coordinates all day, I ended up trying things on semi-random to get things in place...

prairiesnpr commented:
> Getting the coordinates right on the rotated image is quite a struggle.. I think some kind of documentation of intended workflow would be good. I have now set something up - will know more on how good it is tomorrow when the mower gets going again, but to be honest even taking help from my wife who works with maps and coordinates all day, I ended up trying things on semi-random to get things in place...
> 
> ![mover_sample](https://user-images.githubusercontent.com/58572670/246526845-adddb345-9c65-44c5-9df7-8a905af9b0c8.png)

Thanks for sharing the Pic, always awesome to see something you wrote in use.

prairiesnpr commented:
@Din-BH Here is an initial detailed guide on how to create an image and configure the camera and zones.  Would appreciate any feedback you have on it. https://github.com/prairiesnpr/husqvarna_automower/blob/dev/MAP_GUIDE.md

Din-BH commented:
> @Din-BH Here is an initial detailed guide on how to create an image and configure the camera and zones. Would appreciate any feedback you have on it. https://github.com/prairiesnpr/husqvarna_automower/blob/dev/MAP_GUIDE.md

This is very very good! In my case I was working with an image taken from a quad copter on high altitude, which makes it require more tinkering, but from here most should be able to make the pic to their liking I think.

The fine tuning is still complicated. When you get almost the right fit to the image but you find you should 'pinch zoom' it a bit to make perfect, working with the coordinate system rotated from middle of image coordinates makes changes seem random, particularly when the rotation is big - in my case -141.

Din-BH commented:
Still would suggest to take out the extra save step. Since the change is not visible until you save, you actually want the save to happen as soon as possible so you can see the result.

Thomas55555 commented:
Can this file be deleted: 
`husqvarna_automower/custom_components/__init__.py` ?

prairiesnpr commented:
> Can this file be deleted: `husqvarna_automower/custom_components/__init__.py` ?

No, had to add it for test discovery to work.

prairiesnpr commented:
@Thomas55555 Take a look at the last [commit](https://github.com/Thomas55555/husqvarna_automower/pull/485/commits/7dc432068a79a56c4e95c7cd3744f4db5fc68e18),this will raise ConfigEntryAuthFailed if data isn't returned from the api.  Not certain I like this though, feel like the error should really originate from aioautomower, but this should allow the user to reauth or just reload if it happens again.

prairiesnpr commented:
This should address all the comments to this point, we now retain the existing mower selections and pass them to the form.

Thomas55555 commented:
> 

Please leave it as it was before the PR, because we are not knowing where it's coming from. This may cause unneeded reauths, which we don't want.

prairiesnpr commented:
> > @Thomas55555 Take a look at the last [commit](https://github.com/Thomas55555/husqvarna_automower/pull/485/commits/7dc432068a79a56c4e95c7cd3744f4db5fc68e18),this will raise ConfigEntryAuthFailed if data isn't returned from the api. Not certain I like this though, feel like the error should really originate from aioautomower, but this should allow the user to reauth or just reload if it happens again.
> 
> Please leave it as it was before the PR, because we are not knowing where it's coming from. This may cause unneeded reauths, which we don't want.

We could reload the entry?  I hate to just leave the integration in a broken state.  Another thought might be, that aioautomower should return the last valid data and log the error, if it is a transient issue with the api, it might be resolved the next time we poll the api.

Thomas55555 commented:
Maybe it was just a coincidence with the invalid credentials issue. And polling again didn't take effect. There was no new data for over 24h.
We can release as beta first, test it and add fixes, if it's really related to that.

prairiesnpr commented:
OK, reverted, we no longer raise ConfigEntryAuthFailed if the API returns a 400 error.

Thomas55555 commented:
Okay, I've tested it now.
It's much better, with saving zones automatically and remembering assignment of the zones to the mower.
Do we still need the button `Save Zone` now? This could be removed, isn't it?

prairiesnpr commented:
> Okay, I've tested it now. It's much better, with saving zones automatically and remembering assignment of the zones to the mower. Do we still need the button `Save Zone` now? This could be removed, isn't it?

We still need something to signal that we are done and end the flow.  I think best to keep it, otherwise after we saved a zone we would need to end the flow and start the configuration flow over, could maybe rename it to something else if that helps.

Thomas55555 commented:
I just clicked on `submit` and the flow was ended. I think this is sufficient.

prairiesnpr commented:
> I just clicked on `submit` and the flow was ended. I think this is sufficient.

Not sure I'm following you here, are we good as is?

Din-BH commented:
> > I just clicked on `submit` and the flow was ended. I think this is sufficient.
> 
> Not sure I'm following you here, are we good as is?

I for one think the extra save zone is just confusing. The zone is already saved. If you need an entry, I suggest you name it "End configuration" and put it last. If you can do without it, even better.

After moving to the latest dev version, trying to update my Mower in config gives me `User input malformed: Not a list for dictionary value @ data['additional_mowers']`. Also noticed that I can't find a way to add more mowers anywhere - not that I need that, but I guess that was kind of a point to support multiple?

prairiesnpr commented:
> 
> After moving to the latest dev version, trying to update my Mower in config gives me `User input malformed: Not a list for dictionary value @ data['additional_mowers']`. Also noticed that I can't find a way to add more mowers anywhere - not that I need that, but I guess that was kind of a point to support multiple?

I had tried to remove it earlier and ended up where submit would repeatedly reload the window if nothing was selected, since there was no user input it thought we were loading it for the first time. Thought that was super confusing, there's probably a way to remove it, it might come to me.

I changed the way we store data, in order to allow us to prefill the multiple choice lists.  There was no way to do that with the select helper, had to move to cv.multiple which does support default values. The problem though is this breaks existing config for those of us testing.  

Coincidentally, your second question is why you are seeing this, when you set up the camera the last choice is to select additional mowers that should be shown on the same map, since we only have a single mower, the list is empty and not displayed, since we can't show the same mower twice.  But since I changed the multiselect, we went from passing the data as a list of dicts to a dict. Think this is the cause of your error, your existing options dict has an empty list, it expects an empty dict. Long story short, delete the integration via the ui then reinstall it and you should be good, that way you will have your options in the correct format. 

I didn't bother migrating or catching it, since it applies only to those of us testing, same with the migration, if you had an earlier commit, the migration logic will fail since it expects to migrate from a pre PR config. If that doesn't fix it, please let me know and I'll dig some more. I just opened my dev box which is running the latest commit and had no issues changing the rotation several times.

Din-BH commented:
> > After moving to the latest dev version, trying to update my Mower in config gives me `User input malformed: Not a list for dictionary value @ data['additional_mowers']`. Also noticed that I can't find a way to add more mowers anywhere - not that I need that, but I guess that was kind of a point to support multiple?
> I didn't bother migrating or catching it, since it applies only to those of us testing, same with the migration, if you had an earlier commit, the migration logic will fail since it expects to migrate from a pre PR config. If that doesn't fix it, please let me know and I'll dig some more. I just opened my dev box which is running the latest commit and had no issues changing the rotation several times.

Very good. Just wanted to call it out in case it wasn't this. I thought I might revert to the previous known good when we start to get ready to migrate back in to main, and verify the migration works.

prairiesnpr commented:
> Very good. Just wanted to call it out in case it wasn't this. I thought I might revert to the previous known good when we start to get ready to migrate back in to main, and verify the migration works.

Great, appreciate all the support in testing.  When you write something, you know how it's supposed to work and will think that it's pretty solid until you get other to use it, always amazed at how many things you find that were missed when other people try to use it.  Better to discover issue here then when it's an issue in production and I'm getting pinged on a support incident.  

Also, side note, I'm very envious of your drone imagery, that looks amazing.  Unfortunately, can't justify a drone to make my home assistant look better....

prairiesnpr commented:
[c60747f](https://github.com/Thomas55555/husqvarna_automower/pull/485/commits/c60747f51a8233e89c28a193c68485360ce92836) addresses comments on the zone menu, still don't have a way to get rid of the option, but this does make it clear that it's a closing step, enables translations for those two options, and ensures that the default option is to finish and close.

Thomas55555 commented:
Can I merge it?

Din-BH commented:
> > Can I merge it?
> 
> I'm good with merging, since @Din-BH has been instrumental in helping to test it, let's give him a shot to test the latest changes and if he's good merge it. Hopefully we can get some help with the missing translations, we added a bunch of keys that will need to be translated once we merge.

Works like a charm. ;) I say go! Great work!
